arangod/Aql/OptimizerRules.cpp

```diff
@@ -1943,101 +1943,107 @@ void arangodb::aql::specializeCollectRule(Optimizer* opt,
+          CollectOptions::CollectMethod::kHash);
+
+      if (ExecutionNode const* loop = collectNode->getLoop();
+          loop != nullptr && loop->getLoop() == nullptr &&
+          (loop->getType() == EN::ENUMERATE_LIST ||
+           loop->getType() == EN::TRAVERSAL ||
+           loop->getType() == EN::ENUMERATE_PATHS ||
+           loop->getType() == EN::SHORTEST_PATH)) {
+        // if the COLLECT is contained inside a single loop, and the loop is an
+        // enumeration over an array (in contrast to an enumeration over a
+        // collection/view, then prefer the hashed collect variant. this is
+        // because the loop output is unlikely to be sorted in any way.
+        preferHashCollect = true;
+      }
+
+      if (preferHashCollect) {
         // user has explicitly asked for hash method
         // specialize existing the CollectNode so it will become a
         // HashedCollectBlock later. additionally, add a SortNode BEHIND the
-        // CollectNode (to sort the final result)
-        collectNode->aggregationMethod(CollectOptions::CollectMethod::HASH);
-        collectNode->specialized();
-
-        if (!collectNode->isDistinctCommand()) {
-          // add the post-SORT
-          SortElementVector sortElements;
-          for (auto const& v : collectNode->groupVariables()) {
-            sortElements.emplace_back(v.outVar, true);
-          }
+        // CollectNode (to sort the final result).
+        // this is an in-place modification of the plan.
+        // we don't need to create an additional plan for this.
+        collectNode->aggregationMethod(CollectOptions::CollectMethod::kHash);
 
-          auto sortNode = plan->createNode<SortNode>(plan.get(), plan->nextId(),
-                                                     sortElements, false);
+        // add the post-SORT
+        SortElementVector sortElements;
+        for (auto const& v : collectNode->groupVariables()) {
+          sortElements.emplace_back(v.outVar, true);
+        }
 
-          TRI_ASSERT(collectNode->hasParent());
-          auto parent = collectNode->getFirstParent();
-          TRI_ASSERT(parent != nullptr);
+        auto sortNode = plan->createNode<SortNode>(plan.get(), plan->nextId(),
+                                                   sortElements, false);
 
-          sortNode->addDependency(collectNode);
-          parent->replaceDependency(collectNode, sortNode);
-        }
+        TRI_ASSERT(collectNode->hasParent());
+        auto parent = collectNode->getFirstParent();
+        TRI_ASSERT(parent != nullptr);
+
+        sortNode->addDependency(collectNode);
+        parent->replaceDependency(collectNode, sortNode);
 
         modified = true;
         continue;
       }
 
-      // create a new plan with the adjusted COLLECT node
-      std::unique_ptr<ExecutionPlan> newPlan(plan->clone());
+      // are we allowed to generate additional plans?
+      if (!opt->runOnlyRequiredRules(1)) {
```

Comment:
This sounds (comment) like the method could get a better fitting name.

Reply:
Did not rename the method, but removed the potentially confusing `1` parameter.